File: Maximal_Square.cpp
/**
*NOTE:
*     Time Limit Exceeded when testing 9/10 test.
*     I cant solve this problem. Maybe use DP, think later.
*/
class Solution {
public:
    /**
     * @param matrix: a matrix of 0 and 1
     * @return: an integer
     */
    int maxSquare(vector<vector<int> > &matrix) {
        // write your code here
        int rows = matrix.size();
        int cols = matrix[0].size();
        int maxLen = 0;
        vector<int> x;
        vector<int> y;
        for(int row = 0; row < rows; ++row){
            for(int col = 0; col < cols; ++col){
                if(!matrix[row][col]){
                    continue;
                }else{
                    maxLen = maxLen < 1 ? 1 : maxLen;
                    if(row < rows - 1 && col < cols - 1 ){
                        x.emplace_back(row);
                        y.emplace_back(col+1);
                        x.emplace_back(row+1);
                        y.emplace_back(col+1);
                        x.emplace_back(row+1);
                        y.emplace_back(col);
                        int thisLen = 1;
                        checkNext(thisLen, x, y, matrix, rows, cols);
                        maxLen = maxLen < thisLen ? thisLen : maxLen;
                        x.clear();
                        y.clear();                        
                    }

                }
            }
        }
        return maxLen * maxLen;
    }
    bool checkNext(int& thisLen, vector<int>& x, vector<int>& y, const vector<vector<int> > &matrix, const int& rows, const int& cols){
        vector<int> tx = move(x);
        vector<int> ty = move(y);
        x.clear();
        y.clear();
        int i = 0;
        while(i < tx.size() / 2 + 1){
            if(!matrix[tx[i]][ty[i]]) return false;
            if(ty[i] < cols - 1){
                x.emplace_back(tx[i]);
                y.emplace_back(ty[i] + 1);   
            }
            ++i;
        }
        --i;
        if(tx[i] < rows - 1 && ty[i] < cols - 1){
            x.emplace_back(tx[i] + 1);
            y.emplace_back(ty[i] + 1);
            x.emplace_back(tx[i] + 1);
            y.emplace_back(ty[i]);
        }
        ++i;
        while(i < tx.size()){
            if(!matrix[tx[i]][ty[i]]) return false;
            if(tx[i] < rows - 1){
                x.emplace_back(tx[i] + 1);
                y.emplace_back(ty[i]);
            }
            ++i;            
        }
        ++thisLen;
        if(x.size() == tx.size() + 2){
            checkNext(thisLen, x, y, matrix, rows, cols);
        }
        return false;
        
    }
};

File: Maximal Square.py
class Solution:
    #param matrix: a matrix of 0 and 1
    #return: an integer
    def maxSquare(self, matrix):
        # write your code here
        
        # dp[i][j] = min(dp[i - 1][j - ], up[i], lef) + 1 if matrix[i][j] == 1
        # up[i] = up[i - 1] + 1 if matrix[][i - 1] == 0 else 0
        # left[i] = left[i - 1] + 1 if
        if not matrix:
            return 
        length = len(matrix)
        n = len(matrix[0])
        up = [0] * n
        dp = [[0] * n for __ in range(2)]
        ans = 0
        for i in range(n):
            if matrix[0][i] == 1:
                dp[0][i] = 1 
                up[i] = 1
                ans = max(ans, dp[0][i])

        for i in range(1, length):
            left = 0
            # j == 0
            if matrix[i][0] == 1:
                dp[i % 2][0] = 1
                left = 1
                up[0] += 1
            else:
                up[0] = 0
            for j in range(1, n):
                if matrix[i][j] == 1:
                    dp[i % 2][j] = min(dp[(i - 1) % 2][j - 1], up[j], left) + 1
                    up[j] += 1
                    left += 1
                    ans = max(dp[i % 2][j], ans)
                else:
                    dp[i % 2][j] = 0
                    up[j] = 0
                    left = 0
        # m, n = len(matrix), len(matrix[0])
        # dp = [[0] * n for __ in range(2)]
        # ans = 0
        # for i in range(n):
        #     dp[0][i] = 1 if matrix[0][i] == 1 else 0
        #     ans = max(ans, dp[0][i])
        # for i in range(1, m):
        #     dp[i % 2][0] = 1 if matrix[i][0] == 1 else 0
        #     for j in range(1, n):
        #         if matrix[i][j] == 1:
        #             dp[i % 2][j] = min(dp[(i - 1) % 2][j - 1], dp[i % 2][j - 1], dp[(i - 1) % 2][j]) + 1
        #         else:
        #             dp[i % 2][j] = 0
        #     ans = max(ans, max(dp[i % 2]))
        return ans * ans
        